pgmpy/inference/Sampling.py

```diff
@@ -0,0 +1,196 @@
+        >>> inference = BayesianModelSampling(student)
+        >>> evidence = {'diff': State(var='diff', state=0)}
+        >>> inference.rejection_sample(evidence, 2)
+                intel       diff       grade
+        0  (intel, 0)  (diff, 0)  (grade, 1)
+        1  (intel, 0)  (diff, 0)  (grade, 1)
+        """
+        if evidence is None:
+            return self.forward_sample(size)
+        sampled = DataFrame(columns=self.topological_order)
+        prob = 1
+        while len(sampled) < size:
+            _size = int((size - len(sampled)) / prob)
+            _sampled = self.forward_sample(_size)
+            for i in range(_size):
+                for var in evidence:
+                    if evidence[var] != _sampled[var][i]:
+                        _sampled.drop(i, inplace=True)
+                        break
+            prob = max(len(_sampled) / _size, 0.01)  # 0.01 assumed if len(sampled) is small or zero
+            sampled = sampled.append(_sampled)
+        sampled.reset_index(inplace=True, drop=True)
+        return sampled[:size]
+
+    def likelihood_weighted_sample(self, evidence=None, size=1):
+        """
+        Generates weighted sample(s) from joint distribution of the bayesian
+        network, that comply with the given evidence.
+        'Probabilistic Graphical Model Principles and Techniques', Koller and
+        Friedman, Algorithm 12.2 pp 493.
+
+        Parameters
+        ----------
+        evidence: dict
+            a dict key, value pair as {var: state_of_var_observed}
+            None if no evidence
+        size: int
+            size of sample to be generated
+
+        Returns
+        -------
+        sampled: pandas.DataFrame
+            the generated samples with corresponding weights
+
+        Examples
+        --------
+        >>> from pgmpy.factors.Factor import State
+        >>> from pgmpy.models.BayesianModel import BayesianModel
+        >>> from pgmpy.factors.CPD import TabularCPD
+        >>> from pgmpy.inference.Sampling import BayesianModelSampling
+        >>> student = BayesianModel([('diff', 'grade'), ('intel', 'grade')])
+        >>> cpd_d = TabularCPD('diff', 2, [[0.6], [0.4]])
+        >>> cpd_i = TabularCPD('intel', 2, [[0.7], [0.3]])
+        >>> cpd_g = TabularCPD('grade', 3, [[0.3, 0.05, 0.9, 0.5], [0.4, 0.25,
+        ...         0.08, 0.3], [0.3, 0.7, 0.02, 0.2]],
+        ...         ['intel', 'diff'], [2, 2])
+        >>> student.add_cpds(cpd_d, cpd_i, cpd_g)
+        >>> inference = BayesianModelSampling(student)
+        >>> evidence = {'diff': State(var='diff',state=0)}
+        >>> inference.likelihood_weighted_sample(evidence, 2)
+                intel       diff       grade  _weight
+        0  (intel, 0)  (diff, 0)  (grade, 1)      0.6
+        1  (intel, 1)  (diff, 0)  (grade, 1)      0.6
+        """
+        sampled = DataFrame(index=range(size), columns=self.topological_order)
+        sampled['_weight'] = [1] * size
```

Comment:
Don't use list over here. Instead you could create a ```numpy.array``` of ones using ```np.ones((1, size))```. It is much more memory efficient.